plugin/trino-postgresql/src/test/java/io/trino/plugin/postgresql/TestJoinReorderingWithJoinPushdown.java

```diff
@@ -0,0 +1,140 @@
+import org.testng.annotations.Test;
+
+import java.util.List;
+import java.util.Map;
+import java.util.Optional;
+
+import static io.trino.plugin.jdbc.JdbcJoinPushdownSessionProperties.JOIN_PUSHDOWN_STRATEGY;
+import static io.trino.plugin.jdbc.JdbcMetadataSessionProperties.JOIN_PUSHDOWN_ENABLED;
+import static io.trino.plugin.postgresql.PostgreSqlQueryRunner.createPostgreSqlQueryRunner;
+import static io.trino.sql.planner.assertions.PlanMatchPattern.anyTree;
+import static io.trino.sql.planner.assertions.PlanMatchPattern.equiJoinClause;
+import static io.trino.sql.planner.assertions.PlanMatchPattern.join;
+import static io.trino.sql.planner.assertions.PlanMatchPattern.node;
+import static io.trino.sql.planner.assertions.PlanMatchPattern.tableScan;
+import static io.trino.sql.planner.plan.JoinNode.DistributionType.PARTITIONED;
+import static io.trino.sql.planner.plan.JoinNode.Type.INNER;
+import static io.trino.tpch.TpchTable.CUSTOMER;
+import static io.trino.tpch.TpchTable.NATION;
+import static org.assertj.core.api.Assertions.assertThat;
+
+public class TestJoinReorderingWithJoinPushdown
+        extends AbstractTestQueryFramework
+{
+    @Override
+    protected QueryRunner createQueryRunner()
+            throws Exception
+    {
+        TestingPostgreSqlServer postgreSqlServer = closeAfterClass(new TestingPostgreSqlServer());
+        DistributedQueryRunner distributedQueryRunner = createPostgreSqlQueryRunner(
+                postgreSqlServer,
+                Map.of(),
+                Map.of(),
+                List.of(CUSTOMER, NATION));
+
+        postgreSqlServer.execute("ANALYZE " + CUSTOMER.getTableName());
+        postgreSqlServer.execute("ANALYZE " + NATION.getTableName());
+
+        return distributedQueryRunner;
+    }
+
+    @Test
+    public void testJoinReordering()
+    {
+        // disable join pushdown in default session
+        // also disable dynamic filtering to simplify plans for ease of matching
+        Session session = Session.builder(getSession())
+                .setSystemProperty("enable_dynamic_filtering", "false")
+                .setCatalogSessionProperty(getSession().getCatalog().orElseThrow(), JOIN_PUSHDOWN_ENABLED, "false")
+                .build();
+        Session joinPushdownEnabled = Session.builder(session)
+                .setCatalogSessionProperty(session.getCatalog().orElseThrow(), JOIN_PUSHDOWN_ENABLED, "true")
+                // force a join pushdown if possible otherwise join may not be pushed down depending on statistics
+                .setCatalogSessionProperty(session.getCatalog().orElseThrow(), JOIN_PUSHDOWN_STRATEGY, "EAGER")
+                .build();
+
+        String sql = "SELECT c.name, o.clerk, n.name " +
+                "FROM customer c " +
+                "INNER JOIN nation n ON c.nationkey = n.nationkey " +
+                "INNER JOIN tpch.tiny.orders o ON c.custkey = o.custkey";
+
+        PlanMatchPattern joinWithoutFilter =
+                anyTree(
```

Comment:
newbie question: How should I read this plan?

`customer` & `nation` get joined and their outcome gets joined with `orders`?

I'd be happy to have a few comments explaining what triggers the reordering of the join operations.

Reply:
For join re-ordering logic see `ReorderJoins`. This test however doesn't care how joins get re-ordered. It just cares that for a known-suboptimal order joins get reordered even if there's possibility of join pushdown. i.e. we shouldn't eagerly pushdown things because it's possible join re-ordering will lead to better plan overall.